linera-execution/src/system.rs

```diff
@@ -517,9 +517,36 @@ where
                 for (chain_id, stream_id, next_index) in streams {
                     let subscriptions = self
                         .event_subscriptions
-                        .get_mut_or_default(&(chain_id, stream_id))
+                        .get_mut_or_default(&(chain_id, stream_id.clone()))
                         .await?;
-                    subscriptions.next_index = subscriptions.next_index.max(next_index);
+                    if subscriptions.next_index >= next_index {
+                        continue;
+                    }
+                    for application_id in &subscriptions.applications {
+                        txn_tracker.add_stream_to_process(
+                            *application_id,
+                            chain_id,
+                            stream_id.clone(),
+                            subscriptions.next_index,
+                            next_index,
+                        );
+                    }
+                    subscriptions.next_index = next_index;
+                    let index = next_index
+                        .checked_sub(1)
+                        .ok_or(ArithmeticError::Underflow)?;
+                    let event_id = EventId {
+                        chain_id,
+                        stream_id,
+                        index,
+                    };
+                    ensure!(
+                        self.context()
+                            .extra()
+                            .contains_event(event_id.clone())
+                            .await?,
+                        ExecutionError::EventNotFound(event_id)
+                    );
```

Comment:
Wouldn't it make more sense to first check if we have the event, even before we add it to the txn_tracker?

Reply:
It doesn't matter because if this fails, the transaction tracker is dropped anyway. And I can't easily reorder this because `self` is borrowed. I'd have to clone the whole `subscriptions.applications` collection.